### PR TITLE
Sockets - Rosalyn

### PR DESCRIPTION
**Pull request by ranuseh**
# JS Adagrams
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
--- | ---
What patterns were you able to use from your Ruby knowledge to apply to JavaScript? | If statement were actually very similar. With the exception of the addition of curly braces it was very similar in structure. Access and assigning to a hash was also very similar to ruby. Some methods like split. concat, length was also easily transferrable.
Did you need to use different strategies to find information online about JavaScript than you would for Ruby? | Not really, the same sources that I typically use for Ruby (Dev Docs and Stack Overflow) was available for JS, in some cases there is more information about it.
What was a challenge you were able to overcome on this assignment? | Finding 10 random numbers was a big challenge. In ruby you could easily use .random(10) but JS was way more convoluted. It required using some other method (math.random) and essentially hacking it to produce a number for you which i then used as an index to find a number.
What has been interesting and positive about learning a new programming language? | It's interested how quickly I have picked this up. While on the surface it feels very strange, when you get down to it, it works in very much the same way as Ruby.
What is something to focus on your learnings in JavaScript in the next week? | A better understanding for the for loops and the different ways to write a function and why you would choose one style over another.

### CODE REVIEW

File: src/adagrams.js [hunk 1]
```diff
@@ -1,7 +1,137 @@
+import { worker } from "@jest/reporters/build/coverage_worker";
+
+const alphaLetters = {
+  A: 9,
+  N: 6,
+  B: 2,
```
Comment (droberts-sea) [inline src/adagrams.js:6]: While I really like this choice of datastructure, I feel like this would be easier to read if the letters were in order.

File: src/adagrams.js [hunk 2]
```diff
@@ -1,7 +1,137 @@
...
+  L: 1,
+  Y: 4,
+  M: 3,
+  Z: 10
+};
+
+const createLetters = function(alphaLetters) {
+  let lettersArray = [];
+  for (let key in alphaLetters) {
+    const result = key.repeat(alphaLetters[key]).split("");
```
Comment (droberts-sea) [inline src/adagrams.js:64]: I really like that you store the letter frequencies and then build the pool on the fly - this is much easier to read than a giant array full of 9 'A's, 2 'B's, etc.

File: src/adagrams.js [hunk 3]
```diff
@@ -1,7 +1,137 @@
...
+    lettersArray = lettersArray.concat(result);
+  }
+  return lettersArray;
+};
+
 const Adagrams = {
   drawLetters() {
-    // Implement this method for wave 1
+    const allLetters = createLetters(alphaLetters);
+    const letterCopy = allLetters.slice(0);
```
Comment (droberts-sea) [inline src/adagrams.js:73]: Since you create a fresh letter pool every time this function is called, you don't need to make a copy of it on line 73.

**Comment by droberts-sea** [comment]
# JS Adagrams
## What We're Looking For



Feature | Feedback
--- | ---
**General** |
Answered comprehension questions | yes
Small commits with meaningful commit messages | I would like to see much more frequent commits. I know it can be hard to do this when you're also trying to spin up on a new language, but I would argue that makes having that history all the more valuable.
**Code Requirements** |
`drawLetters` method | yes
Uses appropriate data structure to store the letter distribution | yes
All tests for `drawLetters` pass | yes
`usesAvailableLetters` method | yes
All tests for `usesAvailableLetters` pass | yes
`scoreWord` method | yes
Uses appropriate data structure to store the letter scores | yes
All tests for `scoreWord` pass | yes
`highestScoreFrom` method | yes
Appropriately handles edge cases for tie-breaking logic | yes
All tests for `highestScoreFrom` pass | yes
**Overall** |Great job overall! This code is generally clean and well-organized, and it feels like the JavaScript unit is off to a solid start. Keep up the hard work!
